src/main/java/com/linecorp/armeria/server/http/encoding/HttpEncodedResponse.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2016 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.http.encoding;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.List;
+import java.util.zip.DeflaterOutputStream;
+
+import javax.annotation.Nullable;
+
+import org.reactivestreams.Subscriber;
+
+import com.google.common.net.MediaType;
+
+import com.linecorp.armeria.common.http.HttpData;
+import com.linecorp.armeria.common.http.HttpHeaderNames;
+import com.linecorp.armeria.common.http.HttpHeaders;
+import com.linecorp.armeria.common.http.HttpObject;
+import com.linecorp.armeria.common.http.HttpResponse;
+import com.linecorp.armeria.common.http.HttpStatusClass;
+import com.linecorp.armeria.common.stream.FilteredStreamMessage;
+import com.linecorp.armeria.common.stream.StreamMessage;
+
+import io.netty.handler.codec.compression.ZlibWrapper;
+
+/**
+ * A {@link FilteredStreamMessage} that applies HTTP encoding to {@link HttpObject}s as they are published.
+ */
+class HttpEncodedResponse extends FilteredStreamMessage<HttpObject, HttpObject> implements HttpResponse {
+
+    private final ZlibWrapper zlibWrapper;
+    private final List<MediaType> encodableContentTypes;
+    private final int minBytesToForceChunkedAndEncoding;
+
+    @Nullable
+    private ByteArrayOutputStream encodedStream;
+
+    @Nullable
+    private DeflaterOutputStream encodingStream;
+
+    private boolean headersSent;
+
+    public HttpEncodedResponse(
+            StreamMessage<HttpObject> delegate,
+            ZlibWrapper zlibWrapper,
+            List<MediaType> encodableContentTypes,
+            int minBytesToForceChunkedAndEncoding) {
+        super(delegate);
+        this.zlibWrapper = zlibWrapper;
+        this.encodableContentTypes = encodableContentTypes;
+        this.minBytesToForceChunkedAndEncoding = minBytesToForceChunkedAndEncoding;
```

Comment:
Please do some validation - null checks, `minBytestoForceChunkedAndEncoding` cannot be negative, ..